src/main/java/org/terasology/taskSystem/TaskManagementSystem.java

```diff
@@ -49,94 +59,233 @@
     @In
     private EntityManager entityManager;
 
-    @In
-    private NUIManager nuiManager;
-
     @In
     private Time timer;
 
     private HoldingComponent oreonHolding;
     private String newTaskType;
-    private TaskComponent task;
-    private UIScreenLayer taskSelectionScreenLayer;
+    private TaskComponent taskComponent;
```

Comment:
Not sure if there's really anything wrong with this, it might be a perfectly legitimate design pattern, but historically seeing components stored in a system makes me worry a bit. Maybe less so if it is a template of some sort you clone out as needed, but to track things ... if it also gets used attached to an entity and ever gets attached to more than one entity at a time while also existing here that makes my code smell senses tingle :-)

Reply:
This task component is set to a new instance every time a task creation process is initiated. It is used in the `recieveNewTask` method which sends the event responsible for opening the task selection screen, whose event handler is present in the client system. The client system then sends a event which is handled in the authority in yet another method `recieveSetTaskEvent`. Both the methods in the authority system make changes to the component before adding it to the task entity, and since the Client system comes in the middle of the flow I used a global `taskComponent` variable.

Reply:
Hmm, thanks for the extra details, but my code smell sense is still tingling. I can't be sure without digging way deeper so this isn't guaranteed, but perhaps likely :-)

Looking a little closer before I step away for a bit I'm spotting a few related bits, like the use of `oreonHolding` as well, and the event handler definition for `receiveNewTask` - I suspect if a player has a different module enabled that also uses `ApplyBlockSelectionEvent` then your handler would trigger as well, even if entirely unrelated, and there isn't any conditional logic that checks to see the player was really selecting some blocks to do something for Oreons.

I suspect, but haven't looked deeply enough yet to point out examples line by line, that you could run into problems with a larger ... task landscape? As in, a more complicated village where tasks are being created, claimed, performed, and even abandoned/exchanged at times. Your code _works_ that much is clear, but it looks vulnerable to timing bugs.

Adding in multiplayer later would also get trickier than it would need to be. I wouldn't mind merging this as-is and improving it later, but I really think we need to find an approach that doesn't rely on keeping state in a `System` via orphaned `Components` that end up interacting / getting used by different entities. Maybe more of a brainstorming on Slack sometime when we're both around :-)

An initial "fix" that is still mostly a bandaid would be tracking the Oreon Holding _entity_ at a top level in `TaskManagementSystem` (with the assumption there'll only ever be one) but then solely using it to fetch/update data in its components as needed. That would lower the risk of timing bugs but still require rework later to support multiple holdings or even multiple players interacting with them